### PR TITLE
chore: create internal dev-schematics project

### LINKED ISSUE #7374: Create "schematics dev" project
According to our conversation on schematics and installation script let's create `schematics-dev` (let's try to find a better name) project. This will allow us to distinguish standard/core/base CI-like operations from more specific ones (like cds or testing some features, ie. outlets).

FYI @dunqan @marlass

### PR DESCRIPTION
**Pull request by psmul**
Closes: #7374

### CODE REVIEW

File: projects/dev-schematics/src/add-test-outlets/files/test-outlets/test-outlet.module.ts.template [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { NgModule } from '@angular/core';
```
Comment (marlass) [inline projects/dev-schematics/src/add-test-outlets/files/test-outlets/test-outlet.module.ts.template]: Can we somehow directly pull those files from storefrontapp?
I would love to maintain this in one place. Not in app and schematics.

**Review by marlass — CHANGES_REQUESTED** [review]

File: projects/dev-schematics/src/add-test-outlets/files/test-outlets/test-outlet.module.ts.template (hunk 1 above)
Comment (psmul) [reply projects/dev-schematics/src/add-test-outlets/files/test-outlets/test-outlet.module.ts.template]: Sadly, we can't. Schematics operate on an active directory when applied. Also renaming files to contain '*.template' isn't accidental, because Schematics can't copy / move / do whatever with files if they're not flagged in any way. I will try to dig into this issue once more

Comment (marlass) [reply projects/dev-schematics/src/add-test-outlets/files/test-outlets/test-outlet.module.ts.template]: But could we in schematics build script create those template files dynamically into final bundle from the repo?

File: projects/dev-schematics/package.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  ],
+  "scripts": {
+    "build": "tsc -p tsconfig.json",
+    "test": "tsc -p tsconfig.json && jasmine \"src/**/*_spec.js\""
+  },
+  "author": "",
+  "license": "MIT",
+  "schematics": "./src/collection.json",
+  "dependencies": {
+    "@angular-devkit/core": "^9.1.0",
```
Comment (znikola) [inline projects/dev-schematics/package.json:17]: TODO: can some of these deps be part of `devDependencies`?

File: projects/dev-schematics/src/add-baseSite/index.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    `;
+    insertPropertyInStorefrontModuleCallExpression(
+      tree,
+      'src/app/app.module.ts',
+      insertion
+    );
+  };
+}
+
+export default function () {
```
Comment (znikola) [inline projects/dev-schematics/src/add-baseSite/index.ts]: Can we specify a return type?

File: projects/dev-schematics/src/add-baseSite/index.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { chain, Tree } from '@angular-devkit/schematics';
+import { insertPropertyInStorefrontModuleCallExpression } from '../shared/utils/module-file-utils';
+
+function provideTestBaseSites() {
```
Comment (znikola) [inline projects/dev-schematics/src/add-baseSite/index.ts]: Can we specify a return type?

File: projects/dev-schematics/src/add-routing/index.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { chain, Tree } from '@angular-devkit/schematics';
+import { insertPropertyInStorefrontModuleCallExpression } from '../shared/utils/module-file-utils';
+
+function provideTestRouting() {
```
Comment (znikola) [inline projects/dev-schematics/src/add-routing/index.ts]: Can we specify a return type?

File: projects/dev-schematics/src/add-routing/index.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      }`;
+    insertPropertyInStorefrontModuleCallExpression(
+      tree,
+      'src/app/app.module.ts',
+      insertion
+    );
+  };
+}
+
+export default function () {
```
Comment (znikola) [inline projects/dev-schematics/src/add-routing/index.ts]: Can we specify a return type?

File: projects/dev-schematics/src/add-test-outlets/files/test-outlets/test-outlet.module.ts.template (hunk 1 above)
Comment (znikola) [reply projects/dev-schematics/src/add-test-outlets/files/test-outlets/test-outlet.module.ts.template]: I like this suggestion to write a script for this. Maybe the script can do the following:
- copy these files to a certain path in the dev-schematics projects (this folder should be git-ignored)
- append `.template` to all files that were copied

It can also be very useful for iterative development to clean this directory before copying files.

File: projects/dev-schematics/src/add-test-outlets/index.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      addToModuleImportsAndCommitChanges(tree, modulePath, TEST_OUTLET_MODULE);
+    }
+  };
+}
+
+function provideTestOutletsModuleFiles(): Source {
+  return apply(url('./files'), [renameTemplateFiles(), move('.', './src/app')]);
+}
+
+export default function (options: DevSpartacusOptions) {
```
Comment (znikola) [inline projects/dev-schematics/src/add-test-outlets/index.ts]: Can we specify a return type?

File: projects/dev-schematics/src/ng-add/index.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  noop,
+  Rule,
+  schematic,
+  SchematicContext,
+  SchematicsException,
+  Tree,
+} from '@angular-devkit/schematics';
+import { getSpartacusVersion } from '../shared/utils/package-utils';
+
+export default function (options: any): Rule {
```
Comment (znikola) [inline projects/dev-schematics/src/ng-add/index.ts]: Can we specify a type for `options`? Would `DevSpartacusOptions` work? Or maybe `Schema`?

File: projects/schematics/src/ng-add/index.ts [hunk 9]
```diff
@@ -21,7 +19,6 @@ export default function (options: any): Rule {
       enableSSR && JSON.parse(options.ssr)
         ? schematic('add-ssr', options)
         : noop(),
-      externalSchematic(ANGULAR_LOCALIZE, 'ng-add', options),
```
Comment (znikola) [inline projects/schematics/src/ng-add/index.ts]: Was this file changed on purpose?

File: projects/dev-schematics/src/shared/utils/package-utils.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import { Tree } from '@angular-devkit/schematics';
+import { SPARTACUS_CORE } from '../constants';
+
+export function getSpartacusVersion(tree: Tree): string {
+  const buffer = tree.read('package.json');
+  let spartacusVersion = '';
+  if (buffer) {
+    const packageJson = JSON.parse(buffer.toString('utf-8'));
```
Comment (znikola) [inline projects/dev-schematics/src/shared/utils/package-utils.ts]: We can use the constant here ;)

File: projects/dev-schematics/src/shared/utils/module-file-utils.ts [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+  moduleSource = moduleSource || getTsSourceFile(host, modulePath);
+  return addSymbolToNgModuleMetadata(
+    moduleSource,
+    modulePath,
+    metadataType,
+    importText
+  ) as InsertChange[];
+}
+
+export function insertPropertyInStorefrontModuleCallExpression(
```
Comment (znikola) [inline projects/dev-schematics/src/shared/utils/module-file-utils.ts:15]: Can we specify the return type?

**Review by znikola — CHANGES_REQUESTED** [review]

File: projects/dev-schematics/README.md [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+## Adding dev-schematics to your Spartacus project
+
+Run the following command from dev-schematics project root:
+
+`yarn build && npm pack`
+
+Run the following command from your project root:
+
+`ng add <path to your built spartacus-dev-schematics-x.x.x.tgz file>`
```
Comment (znikola) [inline projects/dev-schematics/README.md]: We can maybe add a link to the `projects/schematics/README.md` docs and point to another of testing this by using verdaccio?
This is just for the sake of consistency.

For this to work, we can improve the `scripts/migrations-test.sh` by adding an argument (e.g. `dev`). When this argument is passed, the script should build and publish only _dev-schematics_ project.

This approach is also useful when we make a change to both schematics and dev-schematics and want to test both changes at the same.

File: projects/dev-schematics/README.md [hunk 13]
```diff
@@ -0,0 +1,19 @@
+# Getting Started
```
Comment (znikola) [inline projects/dev-schematics/README.md:1]: Let's document a bit more on how to use it.

1. `ng new ...`
2. `ng add @spartacus/schematics ....` . Should it be 1.5+ or 2.0+?
3. `ng add <path>`
4. some example command

File: projects/dev-schematics/scripts/outlet-template-files.sh [hunk 14]
```diff
@@ -0,0 +1,27 @@
+
```
Comment (znikola) [inline projects/dev-schematics/scripts/outlet-template-files.sh:1]: Can we document how to use this script in the `README.md`?
A few words on what it does, and a few examples would be nice.

File: projects/dev-schematics/README.md (hunk 13 above)
Comment (znikola) [reply projects/dev-schematics/README.md:1]: Some use cases that we can cover:
- using multiple options as part of the same command
- document a case when we already added the `dev-schematics` with `baseSite` option, and now we want to add `routing` for example on top of it.
Mentioning that this is possible would help people save some time.

File: projects/dev-schematics/src/shared/constants.ts [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+/***** Removed public api end *****/
+
+/***** Properties start *****/
+
+/***** Properties end *****/
+
+/***** APIs start *****/
+
+/***** APIs end *****/
```
Comment (znikola) [inline projects/dev-schematics/src/shared/constants.ts]: We can cleanup these sections, as they're not populated

**Review by znikola — CHANGES_REQUESTED** [review]

File: projects/dev-schematics/package.json [hunk 16]
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "@spartacus/dev-schematics",
+  "version": "0.0.1",
+  "description": "Spartacus dev-schematics",
+  "keywords": [
+    "spartacus",
+    "dev-schematics"
+  ],
+  "scripts": {
+    "build": "tsc -p tsconfig.json",
```
Comment (znikola) [inline projects/dev-schematics/package.json]: Let's add a `clean` command and hook it into the `build` and `test`.
For the reference, take a look at the schematics projects' `package.json`

File: projects/dev-schematics/package.json [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+  "license": "MIT",
+  "schematics": "./src/collection.json",
+  "dependencies": {
+    "@angular-devkit/core": "^9.1.0",
+    "@angular-devkit/schematics": "^9.1.0",
+    "@schematics/angular": "^9.1.0",
+    "@types/jasmine": "^3.5.0",
+    "@types/node": "^12.11.1",
+    "jasmine": "^3.5.0",
+    "typescript": "~3.8.3"
```
Comment (znikola) [inline projects/dev-schematics/package.json]: Let's move these to devDependencies section

File: projects/dev-schematics/README.md [hunk 18]
```diff
@@ -1,19 +1,55 @@
...
 ### Available options
 
-- `baseSite`: Adds pre-defined base site list to app module
-- `routing`: Adds pre-defined routing to app module
-- `outlets`: Adds testing outlets module to app module
+Options are provided as flags for main install command as `--<option>`:
+
+- `baseSites`: Adds pre-defined base site list to the app module
+- `routing`: Adds pre-defined routing to the app module
+- `outlets`: Adds testing outlets module to app module. **NOTE:** If you plan to use testing outlets module, please remember to run `outlet-template-files.sh` bash script beforehand, as it will ensure up to date testing module files to be applied.
```
Comment (znikola) [inline projects/dev-schematics/README.md]: Can we run the script automatically? We could maybe use `shelljs` package and execute the script when `outlets` option is specified.

File: projects/dev-schematics/README.md [hunk 19]
```diff
@@ -1,19 +1,55 @@
...
+- `ng g @spartacus/dev-schematics:add-routing` - provides basic routing for testing 
+
+- `ng g @spartacus/dev-schematics:add-test-outlets` - provides modules and files for outlets testing
+
+
+## Example of full dev-schematics implementation for Spartacus project
+
+1. Run bash script from source, ensuring up to date test outlets files
+    - `cd projects/dev-schematics/scripts && ./outlet-template-files.sh`
+1. Build and pack dev-schematics project from source:
```
Comment (znikola) [inline projects/dev-schematics/README.md]: `1` again :)

File: projects/dev-schematics/README.md [hunk 20]
```diff
@@ -1,19 +1,55 @@
...
+
+- `ng g @spartacus/dev-schematics:add-routing` - provides basic routing for testing 
+
+- `ng g @spartacus/dev-schematics:add-test-outlets` - provides modules and files for outlets testing
+
+
+## Example of full dev-schematics implementation for Spartacus project
+
+1. Run bash script from source, ensuring up to date test outlets files
+    - `cd projects/dev-schematics/scripts && ./outlet-template-files.sh`
```
Comment (znikola) [inline projects/dev-schematics/README.md]: As suggested above, maybe we can run the script automatically?

File: projects/dev-schematics/README.md [hunk 21]
```diff
@@ -1,19 +1,55 @@
...
+1. Build and pack dev-schematics project from source:
+    - `cd projects/dev-schematics && yarn build && npm pack`
+2. Create new Angular app (or go to existing one):
```
Comment (znikola) [inline projects/dev-schematics/README.md]: Maybe rephrase it like this: ...`or use an* existing one:`

File: projects/dev-schematics/README.md [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+
+`yarn build && npm pack`
+
+Run the following command from your project root:
+
+`ng add <path to your built spartacus-dev-schematics-x.x.x.tgz file>`
+
+> NOTE: Please be aware that dev-schematics without params is an empty shell that doesn't perform any actions.
+
+Alternative method of installing dev-schematics can be achieved through *verdaccio*, about which you can read more here: [Using verdaccio for local development](https://github.com/SAP/spartacus/tree/develop/projects/schematics#testing-update-schematic)
```
Comment (znikola) [inline projects/dev-schematics/README.md]: As this PR https://github.com/SAP/spartacus/pull/7601/files updates the page, can we point to a section from it?

File: projects/dev-schematics/src/shared/utils/file-utils.ts [hunk 23]
```diff
@@ -0,0 +1,63 @@
+import { SchematicsException, Tree } from '@angular-devkit/schematics';
```
Comment (znikola) [inline projects/dev-schematics/src/shared/utils/file-utils.ts]: After discussing with @tobi-or-not-tobi , we agreed to expose a very minimal `public_api` from _schematics_ project. 
Let's do that in this PR. 
At the same time, we can maybe refactor the method signatures a bit, in order to have one exposed "entry" function, to which we can pass all the required parameters.

File: projects/dev-schematics/src/shared/utils/module-file-utils.ts [hunk 24]
```diff
@@ -0,0 +1,102 @@
+import { Tree } from '@angular-devkit/schematics';
```
Comment (znikola) [inline projects/dev-schematics/src/shared/utils/module-file-utils.ts:1]: The same as for `file-utils`

File: projects/dev-schematics/src/shared/utils/package-utils.ts [hunk 25]
```diff
@@ -0,0 +1,12 @@
+import { Tree } from '@angular-devkit/schematics';
```
Comment (znikola) [inline projects/dev-schematics/src/shared/utils/package-utils.ts:1]: The same as for `file-utils`

File: projects/dev-schematics/README.md (hunk 12 above)
Comment (znikola) [reply projects/dev-schematics/README.md]: Can we improve the `migrations-test.sh` script in this PR?

**Review by znikola — CHANGES_REQUESTED** [review]

File: scripts/migrations-test.sh [hunk 26]
```diff
@@ -54,9 +54,17 @@ doItFor "storefrontlib"
...
 
 cd ../projects/schematics
-unpublish "schematics" && publish "schematics" 
+unpublish "schematics" && publish "schematics"
 cd ../../dist
+
+if [[ "$1" == "dev" ]] || [[ "$2" == "dev" ]]; then
+  cd ../projects/dev-schematics
+  unpublish "dev-schematics"
+  yarn build
```
Comment (znikola) [inline scripts/migrations-test.sh]: Let's do `yarn && yarn build`, in case packages are not yet installed.

File: scripts/migrations-test.sh [hunk 27]
```diff
@@ -1,8 +1,8 @@
 ###########################################################
-# This script builds the relevant spartacus libraries 
-# and publushes them to the local npm registry.
+# This script builds the relevant spartacus libraries
+# and publishes them to the local npm registry.
 # The building part can be skipped by providing the
-# `skip` argument when calling the script:
```
Comment (znikola) [inline scripts/migrations-test.sh:5]: Let's mention that `dev` argument is supported.

File: projects/dev-schematics/src/shared/constants.ts [hunk 28]
```diff
@@ -0,0 +1,13 @@
+/***** Imports start *****/
```
Comment (znikola) [inline projects/dev-schematics/src/shared/constants.ts]: Related to this comment: https://github.com/SAP/spartacus/pull/7554/files#r426770948: we can also expose the whole `constants.ts` file from `schematics` project, in which case we don't need this one. Therefore, we can move `TEST_OUTLET_MODULE` and `TEST_CONFIG_MODULE` constants there.

**Review by znikola — CHANGES_REQUESTED** [review]

File: projects/dev-schematics/README.md (hunk 18 above)
Comment (psmul) [reply projects/dev-schematics/README.md]: Not really. When we run schematics, context is set to currently running bundle / package. IMO there's no option to run it automatically without some hacks.

File: projects/dev-schematics/package.json [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+  "version": "0.0.1",
+  "description": "Spartacus dev-schematics",
+  "keywords": [
+    "spartacus",
+    "dev-schematics"
+  ],
+  "scripts": {
+    "clean": "rm -rf src/**/*.js src/**/*.js.map src/**/*.d.ts",
+    "build": "yarn clean && tsc -p tsconfig.json",
+    "test": "tsc -p tsconfig.json && jasmine \"src/**/*_spec.js\""
```
Comment (znikola) [inline projects/dev-schematics/package.json]: Let's also update this one to: `"test": "yarn build && jasmine \"src/**/*_spec.js\""`

File: projects/dev-schematics/README.md (hunk 18 above)
Comment (znikola) [reply projects/dev-schematics/README.md]: Ah, right.
Maybe we can then:
- add a command to `package.json` that will run the script before building the dev-schematics
- this command can also be used when releasing dev-schematics (if we decide to release in the future)

**Review by znikola — CHANGES_REQUESTED** [review]

File: projects/dev-schematics/src/shared/utils/package-utils.ts [hunk 30]
```diff
@@ -1,5 +1,5 @@
 import { Tree } from '@angular-devkit/schematics';
-import { SPARTACUS_CORE, UTF_8 } from '../constants';
+import { SPARTACUS_CORE, UTF_8 } from '@spartacus/schematics/public_api';
```
Comment (znikola) [inline projects/dev-schematics/src/shared/utils/package-utils.ts]: We shouldn't specify `public_api`, it should be only `@spartacus/schematics`. Does it work without the `public_api` part?

File: projects/dev-schematics/src/shared/utils/module-file-utils.ts [hunk 31]
```diff
@@ -1,62 +1,11 @@
...
-): InsertChange[] {
-  moduleSource = moduleSource || getTsSourceFile(host, modulePath);
-  return addSymbolToNgModuleMetadata(
-    moduleSource,
-    modulePath,
-    metadataType,
-    importText
-  ) as InsertChange[];
-}
+import { ANGULAR_CORE, InsertDirection, getTsSourceFile, commitChanges } from "@spartacus/schematics/public_api";
```
Comment (znikola) [inline projects/dev-schematics/src/shared/utils/module-file-utils.ts]: The same as for `package-utils.ts` - `public_api` part should be omitted.

File: projects/schematics/public_api.d.ts [hunk 32]
```diff
@@ -0,0 +1 @@
+export * from './src/shared';
```
Comment (znikola) [inline projects/schematics/public_api.d.ts]: We shouldn't really be committing this file. We don't have it in other libraries. 
Strange that this one didn't ignore it: https://github.com/SAP/spartacus/pull/7554/commits/e2b851a4f5a99ad9726bf638cf4561631de241d3#diff-1f745621faf1fea6435d185420fce42cR4

I guess it's generated once we run `tsc`?

File: projects/dev-schematics/src/shared/utils/package-utils.ts (hunk 30 above)
Comment (znikola) [inline projects/dev-schematics/src/shared/utils/package-utils.ts]: The same as for other imports - `public_api` should be omitted.

File: projects/dev-schematics/src/shared/utils/module-file-utils.ts (hunk 31 above)
Comment (znikola) [inline projects/dev-schematics/src/shared/utils/module-file-utils.ts]: The same as for other imports - `public_api` should be omitted.

File: projects/dev-schematics/src/shared/constants.ts [hunk 33]
```diff
@@ -1,13 +1,4 @@
-/***** Imports start *****/
-export const SPARTACUS_CORE = '@spartacus/core';
-export const ANGULAR_CORE = '@angular/core';
-/***** Imports end *****/
-
 /***** Classes start *****/
 export const TEST_OUTLET_MODULE = 'TestOutletModule';
```
Comment (znikola) [inline projects/dev-schematics/src/shared/constants.ts]: Let's move these to `schematics` project, so that we can delete this file.

File: projects/dev-schematics/src/add-test-outlets/index.ts [hunk 34]
```diff
@@ -16,16 +16,16 @@ import { Schema as DevSpartacusOptions } from '../ng-add/schema';
...
   TEST_CONFIG_MODULE,
   TEST_OUTLET_MODULE,
 } from '../shared/constants';
 import {
+  SPARTACUS_CORE,
+  getTsSourceFile,
   addImport,
   addToModuleImportsAndCommitChanges,
-} from '../shared/utils/module-file-utils';
+} from '@spartacus/schematics/public_api';
```
Comment (znikola) [inline projects/dev-schematics/src/add-test-outlets/index.ts]: The same as for other imports - `public_api` should be omitted.

**Review by znikola — CHANGES_REQUESTED** [review]

File: projects/dev-schematics/src/ng-add/index.ts [hunk 35]
```diff
@@ -8,27 +8,26 @@ import {
...
       throw new SchematicsException(
         'Could not find Spartacus in package dependencies. Please install Spartacus before using this schematic.'
       );
     }
 
-    context.logger.info(options);
-
     return chain([
-      options['default-base-sites'] && JSON.parse(options['default-base-sites'])
+      options['default-base-sites'] && options['default-base-sites']
```
Comment (znikola) [inline projects/dev-schematics/src/ng-add/index.ts]: Not sure why do we need to check for `options['default-base-sites']` two times?
The same goes for line 26 and 30

**Review by znikola — CHANGES_REQUESTED** [review]

File: projects/dev-schematics/src/shared/utils/package-utils.ts (hunk 30 above)
Comment (psmul) [reply projects/dev-schematics/src/shared/utils/package-utils.ts]: Solved with approach yb adding public_api.js file to "main" field in package.json

File: projects/dev-schematics/scripts/test-dev-schematics-local.sh [hunk 36]
```diff
@@ -0,0 +1,15 @@
...
+if [ $# -eq 0 ]; then
+  echo "ERROR: Spartacus schematics version not provided. Aborting."
+  exit 1
+fi
+
+cd ../../schematics
+npm unpublish @spartacus/schematics --registry http://localhost:4873 --force
+yarn build && npm publish --registry http://localhost:4873
+cd ../dev-schematics
+echo "@spartacus:registry=http://localhost:4873" > .npmrc
```
Comment (znikola) [inline projects/dev-schematics/scripts/test-dev-schematics-local.sh:11]: Nice!

File: projects/dev-schematics/scripts/test-dev-schematics-local.sh [hunk 37]
```diff
@@ -0,0 +1,15 @@
+
+
```
Comment (znikola) [inline projects/dev-schematics/scripts/test-dev-schematics-local.sh]: Let's remove these two lines :)

File: projects/dev-schematics/scripts/test-dev-schematics-local.sh [hunk 38]
```diff
@@ -0,0 +1,15 @@
...
+if [ $# -eq 0 ]; then
```
Comment (znikola) [inline projects/dev-schematics/scripts/test-dev-schematics-local.sh]: I'm wonder if the version is even required?
Personally, I always have one version published to verdaccio, and I don't need to specify it when doing `npm i @spartacus/schematics`.
Not sure about other use cases, maybe other devs have different versions published?

File: projects/dev-schematics/README.md [hunk 39]
```diff
@@ -62,10 +62,10 @@ For ease of usage we provide scripts for smoother installation process:
 There might be situatons, where you want to add new entities to/from public API of schematics, but the @spartacus/schematics hasn't been released yet. You can test it locally by following steps:
 
 1. Launch local verdaccio in yours terminal with command `verdaccio` 
-2. Second step has 2 possible ways:
-    - from project root go to scripts/ and run `./publish-schematics-verdaccio.sh` script, which will automatically do all the work for you with building and publishing your changes to local verdaccio
-    - go to `/project/schematics` and run `yarn build && npm publish --registry http://localhost:4873`
-3. Go to dev-schematics project folder and create new file `.npmrc` with content: `@spartacus:registry=http://localhost:4873`
-4. With verdaccio launched, local schematics published to it, we can install new version for testing: `rm -rf node_modules && yarn`
-    
+2. From project root go to `/projects/schematics` and change lib version in package.json file (for local testing purposes only, please remember to revert it before committing any changes)
```
Comment (znikola) [inline projects/dev-schematics/README.md]: Is there a restriction to which version should we change it? E.g. should the version be greater than the current one?

File: projects/schematics/package.json [hunk 40]
```diff
@@ -46,5 +46,5 @@
       "@spartacus/incubator"
     ]
   },
-  "main": "./public_api.js"
+  "main": "src/public_api.js"
```
Comment (znikola) [inline projects/schematics/package.json:50]: This has to use `.js`?

File: projects/schematics/src/public_api.ts [hunk 41]
```diff
@@ -1,4 +1,4 @@
 /*
  * Public API Surface of schematics
  */
-export * from './src/shared/index';
+export * from './shared';
```
Comment (znikola) [inline projects/schematics/src/public_api.ts]: I think we should add `./shared/index` here?

**Review by znikola — CHANGES_REQUESTED** [review]

File: projects/dev-schematics/scripts/test-dev-schematics-local.sh (hunk 38 above)
Comment (psmul) [reply projects/dev-schematics/scripts/test-dev-schematics-local.sh]: Exactly that's what I thought. It's a simple fallback to let dev decide on the published version.

File: projects/dev-schematics/README.md (hunk 39 above)
Comment (psmul) [reply projects/dev-schematics/README.md]: Well, I believe at first thought I would say Verdaccio works only with bumped version, but on the second one - we're publishing the previous one, so verdaccio isn't aware of other versions to be able to compare with, right?

File: projects/schematics/package.json (hunk 40 above)
Comment (psmul) [reply projects/schematics/package.json:50]: Yes, because when app is built and bundled, it's operating only on .JS files. So we're pointing to already transpiled/bundled file here.

File: projects/dev-schematics/README.md (hunk 39 above)
Comment (znikola) [reply projects/dev-schematics/README.md]: I think the version on the verdaccio should be greater than the version in the testing project.
And it doesn't really matter which version is it, as long as it's greater. If you can verify this, we can then simplify the process and make [this one](https://github.com/SAP/spartacus/pull/7554#discussion_r431727063) optional. So that someone can choose (should they need to), and others don't have to think about it if they only have one version published

File: projects/dev-schematics/scripts/test-dev-schematics-local.sh (hunk 38 above)
Comment (psmul) [reply projects/dev-schematics/scripts/test-dev-schematics-local.sh]: Okay, so I've made argument optional as we can call downgraded versions

File: projects/dev-schematics/README.md (hunk 39 above)
Comment (psmul) [reply projects/dev-schematics/README.md]: Okay, checked. There is no need to provide modified version for lib, as we're republishing the whole lib.

File: tsconfig.json [hunk 42]
```diff
@@ -21,7 +21,8 @@
       "@spartacus/vendor": ["projects/vendor/public_api"],
       "@spartacus/assets": ["projects/assets/src/public_api"],
       "@spartacus/cds": ["projects/cds/public_api"],
-      "@spartacus/incubator": ["projects/incubator/public_api"]
+      "@spartacus/incubator": ["projects/incubator/public_api"],
+      "@spartacus/schematics": ["projects/schematics/public_api"]
```
Comment (znikola) [inline tsconfig.json]: The path should now be `projects/schematics/src/public_api`

File: projects/dev-schematics/README.md [hunk 43]
```diff
@@ -0,0 +1,54 @@
...
+
+> NOTE: Please be aware that dev-schematics without params is an empty shell that doesn't perform any actions.
+
+### Available options
+
+Options are provided as flags for main install command as `--<option>`:
+
+- `baseSites`: Adds pre-defined base site list to the app module
+- `routing`: Adds pre-defined routing to the app module
+- `outlets`: Adds testing outlets module to app module. **NOTE:** If you plan to use testing outlets module, please remember to run `outlet-template-files.sh` bash script beforehand, as it will ensure up to date testing module files to be applied.
```
Comment (znikola) [inline projects/dev-schematics/README.md]: Maybe re-phrase this to sound like a note, instead of a requirement.
As this is done under the hood, it might confuse some people; it would be nice to mention this as well.

**Review by znikola — CHANGES_REQUESTED** [review]

**Review by znikola — DISMISSED** [review]

**Review by znikola — DISMISSED** [review]

File: projects/dev-schematics/.npmignore [hunk 44]
```diff
@@ -0,0 +1,5 @@
+# Ignores TypeScript files, but keeps definitions.
+*.ts
```
Comment (marlass) [inline projects/dev-schematics/.npmignore:2]: Let's ignore spec files as well

File: projects/dev-schematics/tsconfig.json [hunk 45]
```diff
@@ -0,0 +1,26 @@
+{
+  "extends": "../../tslint.json",
```
Comment (marlass) [inline projects/dev-schematics/tsconfig.json]: Is it intenational that we extend in tsconfig tslint file?

File: projects/schematics/src/shared/constants.ts [hunk 46]
```diff
@@ -218,6 +219,9 @@ export const ANONYMOUS_CONSENT_LAUNCH_DIALOG_SERVICE =
   'AnonymousConsentLaunchDialogService';
 export const ANONYMOUS_CONSENT_OPEN_DIALOG_COMPONENT =
   'AnonymousConsentOpenDialogComponent';
+
+export const TEST_OUTLET_MODULE = 'TestOutletModule';
```
Comment (marlass) [inline projects/schematics/src/shared/constants.ts:223]: Where this is used?

File: projects/dev-schematics/.npmignore (hunk 44 above)
Comment (marlass) [reply projects/dev-schematics/.npmignore:2]: Same applies to schematics lib. Why publish the spec files to npm 🤦

**Review by marlass — CHANGES_REQUESTED** [review]

File: projects/dev-schematics/.npmignore (hunk 44 above)
Comment (znikola) [reply projects/dev-schematics/.npmignore:2]: Good point, this should be ignored in both `schematics` and `dev-schematics`. However, we don't plan to publish the `dev-schematics` to npm

File: projects/dev-schematics/tsconfig.json (hunk 45 above)
Comment (psmul) [reply projects/dev-schematics/tsconfig.json]: Oh, small omission, will be fixed with the next commit.

File: projects/dev-schematics/.npmignore (hunk 44 above)
Comment (psmul) [reply projects/dev-schematics/.npmignore:2]: Done

File: projects/schematics/src/shared/constants.ts (hunk 46 above)
Comment (psmul) [reply projects/schematics/src/shared/constants.ts:223]: This one is used in dev-schematics. We simply didn't want to create separate constants.ts just for one line in dev-schematics so we decided to put it in here and include in public api.

File: projects/dev-schematics/.npmignore (hunk 44 above)
Comment (marlass) [reply projects/dev-schematics/.npmignore:2]: For spec files we also have `*_spec.js` and `*_spec.js.map`

**Review by marlass — CHANGES_REQUESTED** [review]

File: projects/dev-schematics/.npmignore [hunk 47]
```diff
@@ -0,0 +1,8 @@
...
+!*.d.ts
+yarn.lock
+.release-it.*
+*_spec.ts
+*_spec.js
+*_spec.map
```
Comment (marlass) [inline projects/dev-schematics/.npmignore]:
```suggestion
*_spec.js.map
```

File: projects/schematics/.npmignore [hunk 48]
```diff
@@ -3,3 +3,6 @@
 !*.d.ts
 yarn.lock
 .release-it.*
+*_spec.ts
+*_spec.js
+*_spec.map
```
Comment (marlass) [inline projects/schematics/.npmignore]:
```suggestion
*_spec.js.map
```

**Review by marlass — CHANGES_REQUESTED** [review]

**Review by marlass — APPROVED** [review]

**Review by znikola — APPROVED** [review]
